crates/shared/src/event_handling.rs

```diff
@@ -69,118 +66,174 @@ where
-            None => self.store.last_event_block().await?,
+    async fn event_block_range(&self) -> Result<(Vec<BlockNumberHash>, bool)> {
+        let current_block = self.block_retriever.current_block().await?;
+        let handled_blocks = if self.last_handled_blocks.is_empty() {
+            // since we don't want `Store` to be responsible for hashes, here we just get
+            // block number and get the `safe` block from it - this is done only on first init
+            let last_handled_block = self.store.last_event_block().await?;
+            self.block_retriever
+                .blocks(RangeInclusive::new(last_handled_block, last_handled_block))
+                .await?
+        } else {
+            self.last_handled_blocks.clone()
         };
-        let current_block = self.block_retriever.current_block_number().await?;
-        let from_block = last_handled_block.saturating_sub(MAX_REORG_BLOCK_COUNT);
-        anyhow::ensure!(
-            from_block <= current_block,
-            format!(
-                "current block number according to node is {} which is more than {} blocks in the \
-                 past compared to last handled block {}",
-                current_block, MAX_REORG_BLOCK_COUNT, last_handled_block
-            )
+
+        let current_block_number = current_block.number.context("missing number")?.as_u64();
+        let current_block_hash = current_block.hash.context("missing hash")?;
+        let (last_handled_block_number, last_handled_block_hash) = *handled_blocks.last().unwrap();
+        tracing::debug!(
+            "current block: {} - {:?}, last_handled_blocks: {} - {:?}",
+            current_block_number,
+            current_block_hash,
+            last_handled_block_number,
+            last_handled_block_hash
         );
-        Ok(BlockNumber::Specific(from_block)..=BlockNumber::Latest(current_block))
+
+        // handle special case which happens most of the time (no reorg, just one new block is added)
+        if current_block.parent_hash == last_handled_block_hash {
+            let is_reorg = false;
+            return Ok((vec![(current_block_number, current_block_hash)], is_reorg));
+        }
+
+        // handle special case when no new block is added
+        if (current_block_number, current_block_hash)
+            == (last_handled_block_number, last_handled_block_hash)
+        {
+            let is_reorg = false;
+            return Ok((vec![], is_reorg));
+        }
+
+        // get all canonical blocks starting from a safe reorg block (last_handled_block_number-25)
+        // and ending with a current canonical block number. This way, we make sure we have
+        // all block numbers and hashes needed for updating the storage
+        let current_blocks = self
+            .block_retriever
+            .blocks(RangeInclusive::new(
+                last_handled_block_number.saturating_sub(MAX_REORG_BLOCK_COUNT),
+                current_block_number,
+            ))
```

Comment:
I think this can be very problematic if we, say, intialize the Balancer registry without the subgraph. The reason is that it will need to fetch block headers for thousands of blocks.

Reply:
You mean if we switch to reindexing the events from scratch? I think in both cases (subgraph or reindexing) we will set/update the initial liquidity pools state on app startup to the block `current_block() - 25` and then proceed with maintenance (btw reindexing will work with `eth_getLogs(from_block, to_block)` instead of `eth_getLogs(block_hash)`).
So, what I can do here is to add a check if a range is too big, then return error? (this would skip the maintenance and raise alerts after some time).

Reply:
OK - I missed that. For some reason I thought it used the same function for indexing from scratch. Can we double check by starting the indexer with:

https://github.com/cowprotocol/services/blob/2350c19744a1ae29d490092e2fe3907e1f57d364/crates/shared/src/sources/balancer_v2/pool_init.rs#L23

Reply:
I double checked and it doesn't work 😞:
```
% cargo test -p shared -- --ignored --nocapture balancer_fetched_pools_match_subgraph
...
2022-09-20T15:42:30.729Z DEBUG shared::transport::http: [][id:3303517] received response: "Request too large."
```

Reply:
> I thought it used the same function for indexing from scratch

Actually it does use the same function. I wanted to revert the code for reindexing from scratch with `eth_getLogs(from_block, to_block)` in the following PR when I implement it for UniswapV3, but was not aware of this UT. :)
I will add that now also.

Reply:
Added hybrid strategy for reindexing from scratch.

Reply:
```
% cargo test -p shared -- --ignored --nocapture balancer_fetched_pools_match_subgraph
... 15 minutes later ...
test sources::balancer_v2::pool_fetching::tests::balancer_fetched_pools_match_subgraph ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 351 filtered out; finished in 1001.43s
```

🎉